### PR TITLE
Melodic polygon performance grid

### PR DESCRIPTION
**Pull request by RainerKuemmerle**
Introduce a grid structure for improving the search for neighboring points

### CODE REVIEW

File: include/costmap_converter/costmap_to_polygons.h [hunk 1]
```diff
@@ -238,14 +264,30 @@ class CostmapToPolygonsDBSMCCH : public BaseCostmapToPolygons
...
    
-   // convex hull parameters
-   double min_keypoint_separation_; //!< Clear keypoints of the convex polygon that are close to each other [distance in meters] (0: keep all)
+   std::vector<std::vector<int> > neighbor_lookup_; //! array of cells for neighbor lookup
+   int neighbor_size_x_;
+   int neighbor_size_y_;
+   double offset_x_;
+   double offset_y_;
+
+   int neighborCellsToIndex(int cx, int cy)
```
Comment (nirwester) [inline include/costmap_converter/costmap_to_polygons.h]: I would add a brief description :)

File: src/costmap_to_lines_convex_hull.cpp [hunk 2]
```diff
@@ -61,18 +61,19 @@ CostmapToLinesDBSMCCH::~CostmapToLinesDBSMCCH()
...
     
-    max_pts_ = 30;
-    nh.param("cluster_max_pts", max_pts_, max_pts_);
+    parameter_.max_pts_ = 30;
+    nh.param("cluster_max_pts", parameter_.max_pts_, parameter_.max_pts_);
     
     // convex hull
-    min_keypoint_separation_ = 0.1;
-    nh.param("convex_hull_min_pt_separation", min_keypoint_separation_, min_keypoint_separation_);
+    parameter_.min_keypoint_separation_ = 0.1;
```
Comment (nirwester) [inline src/costmap_to_lines_convex_hull.cpp]: Can be done on one line puttin 0.1 as default param

File: src/costmap_to_lines_ransac.cpp [hunk 3]
```diff
@@ -60,14 +60,19 @@ CostmapToLinesDBSRANSAC::~CostmapToLinesDBSRANSAC()
 void CostmapToLinesDBSRANSAC::initialize(ros::NodeHandle nh)
 { 
     // DB SCAN
-    max_distance_ = 0.4; 
-    nh.param("cluster_max_distance", max_distance_, max_distance_);
+    parameter_.max_distance_ = 0.4;
```
Comment (nirwester) [inline src/costmap_to_lines_ransac.cpp]: Same as above, the default values can be defined once.

File: test/costmap_polygons.cpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  ASSERT_EQ(neighbors_trivial, neighbors);
+
+  costmap_to_polygons.regionQuery(1, neighbors);
+  regionQueryTrivial(1, neighbors_trivial);
+  ASSERT_EQ(costmap_to_polygons.parameters().max_pts_/2, int(neighbors.size()));
+  ASSERT_EQ(neighbors_trivial.size(), neighbors.size());
+  std::sort(neighbors.begin(), neighbors.end());
+  ASSERT_EQ(neighbors_trivial, neighbors);
+}
+
```
Comment (nirwester) [inline test/costmap_polygons.cpp:107]: Maybe also a test against an empty costmap? :)

File: src/costmap_to_lines_convex_hull.cpp (hunk 2 above)
Comment (RainerKuemmerle) [reply src/costmap_to_lines_convex_hull.cpp]: followed the pattern of the existing code :)

File: include/costmap_converter/costmap_to_polygons.h (hunk 1 above)
Comment (RainerKuemmerle) [reply include/costmap_converter/costmap_to_polygons.h]: sounds reasonable

**Review by nirwester — APPROVED** [review]
